rest-api/offline/public_metrics_export.py

```diff
@@ -0,0 +1,145 @@
+   WHERE race IS NOT NULL
+   GROUP BY 1
+   """,
+   lambda v: Race.lookup_by_number(v).name),
+  ('state',
+   """
+   SELECT SUBSTR(code.value, {}), ps.count
+   FROM (
+    SELECT state_id, SUM(1) count
+    FROM participant_summary
+    WHERE state_id IS NOT NULL
+    GROUP BY 1
+   ) ps LEFT JOIN code
+   ON ps.state_id = code.code_id;
+   """.format(len('PIIState_') + 1),
+   None),
+  ('ageRange',
+   """
+   SELECT
+     CASE
+      WHEN YEARS_OLD[date_of_birth] < 0 THEN 'UNSET'
+      WHEN YEARS_OLD[date_of_birth] <= 17 THEN '0-17'
+      WHEN YEARS_OLD[date_of_birth] <= 25 THEN '18-25'
+      WHEN YEARS_OLD[date_of_birth] <= 35 THEN '26-35'
+      WHEN YEARS_OLD[date_of_birth] <= 45 THEN '36-45'
+      WHEN YEARS_OLD[date_of_birth] <= 55 THEN '46-55'
+      WHEN YEARS_OLD[date_of_birth] <= 65 THEN '56-65'
+      WHEN YEARS_OLD[date_of_birth] <= 75 THEN '66-75'
+      WHEN YEARS_OLD[date_of_birth] <= 85 THEN '76-85'
+      ELSE '86+'
+     END age_range,
+     SUM(1)
+   FROM participant_summary
+   GROUP BY 1;
+   """,
+   None),
+  ('physicalMeasurements',
+   """
+   SELECT physical_measurements_status, SUM(1)
+   FROM participant_summary
+   GROUP BY 1;
+   """,
+   lambda v: PhysicalMeasurementsStatus.lookup_by_number(v).name),
+  ('biospecimenSamples',
+   """
+   SELECT
+     CASE (biospecimen_status)
+      WHEN '{}' THEN 'UNSET'
+      WHEN '{}' THEN 'UNSET'
+      ELSE 'COLLECTED'
+     END, SUM(1)
+   FROM participant_summary
+   GROUP BY 1;
+   """.format(str(OrderStatus.UNSET.number), str(OrderStatus.CREATED.number)),
+   None),
+  _questionnaire_metric('questionnaireOnOverallHealth', 'questionnaire_on_overall_health'),
+  # Personal habits is a newer naming for lifestyle
+  _questionnaire_metric('questionnaireOnPersonalHabits', 'questionnaire_on_lifestyle'),
+  # Sociodemographics is a newer naming for 'the basics'
+  _questionnaire_metric('questionnaireOnSociodemographics', 'questionnaire_on_the_basics'),
+  _questionnaire_metric('questionnaireOnHealthcareAccess', 'questionnaire_on_healthcare_access'),
+  _questionnaire_metric('questionnaireOnMedicalHistory', 'questionnaire_on_medical_history'),
+  _questionnaire_metric('questionnaireOnMedications', 'questionnaire_on_medications'),
+  _questionnaire_metric('questionnaireOnFamilyHealth', 'questionnaire_on_family_health')
+]
+
+
+class PublicMetricsExport(object):
+  """Exports data from the database needed to generate public registration metrics."""
+
+  @staticmethod
+  def export():
+    # TODO(calbach): Write the output to a given target destination rather than
+    # returning it.
+    out = {}
+    # Using a session here should put all following SQL invocations into a
+    # non-locking read transaction per
+    # https://dev.mysql.com/doc/refman/5.7/en/innodb-consistent-read.html
+    with database_factory.make_server_cursor_database().session() as session:
+      for (key, sql, valuef) in _SQL_AGGREGATIONS:
+        sql = replace_years_old(sql)
```

Comment:
On each SQL expression we're going to have to tack on some extra logic that says:

ps.hpo_id != :test_hpoId and ps.email NOT LIKE :test_email_pattern

as shown here:

https://github.com/all-of-us/raw-data-repository/blob/master/rest-api/offline/metrics_export.py#L78

The reason: we have test participants in prod which shouldn't count towards metrics.

And yes, I hate this fact, and no, there isn't anything we can do about it but filter them out. :)